README
Name
    YAML::Safe - Safe Perl YAML Serialization using XS and libyaml

Synopsis
        use YAML::Safe qw(LoadFile DumpFile);

        my $yaml  = Dump [ 1..4 ];
        my $array = Load $yaml;

        $yaml  = DumpFile ("my.yml", [ 1..4 ]);
        $array = LoadFile "my.yml";
        open my $fh, "my.yml";
        $yaml  = DumpFile ($fh, [ 1..4 ]);
        $array = LoadFile $fh;
        open *FH, "my.yml";
        $yaml  = DumpFile (*FH, [ 1..4 ]);
        $array = LoadFile *FH;

        my $obj = YAML::Safe->new;
        $yaml  = $obj->DumpFile ("my.yml", [ 1..4 ]);
        $array = $obj->LoadFile("my.yml");

        $yaml = YAML::Safe->new->nonstrict->encoding("any");
        $yaml->SafeClass("DateTime");
        $array = $yaml->SafeLoadFile("META.yml");

        $yaml = YAML::Safe->new->canonical;
        $yaml->SafeClass("DateTime");
        $array = $yaml->SafeDumpFile("META.yml");

Description
    This module is a refactoring of YAML::XS, the old Perl XS binding to
    "libyaml" which offers Perl somewhat acceptable YAML support to date.
    YAML::XS never produced code which could be read from YAML, and thus was
    unsuitable to be used as YAML replacement for core and CPAN. It also
    required reading and setting options from global variables.

    Kirill Siminov's "libyaml" is a good YAML library implementation. The C
    library is written precisely to the YAML 1.1 specification, and offers
    YAML 1.2 support. It was originally bound to Python and was later bound
    to Ruby. "libsyck" is written a bit more elegant, has less bugs, is not
    as strict as libyaml, but misses some YAML features. It can only do YAML
    1.0

    This module exports the functions "Dump" and "Load", and do work as
    functions exactly like YAML::XS and "YAML.pm"'s corresponding functions.
    It is however preferred to use the new OO-interface to store all options
    in the new created object. YAML::Safe does not support the old globals
    anymore.

    There are also new Safe variants of Load and Dump methods, and options
    as setter methods.

    If you set the option "noindentmap", "YAML::Safe" will behave like with
    version < 0.70, which creates yml files which cannot be read by
    "YAML.pm"

    However the loader is stricter than "YAML", "YAML::Syck" and
    "CPAN::Meta::YAML" i.e. "YAML::Tiny" as used in core. Set the option
    "nonstrict" to allow certain reader errors to pass the "CPAN::Meta"
    validation testsuite.

FUNCTIONS
    Load
    LoadFile
    Dump
    DumpFile
    libyaml_version

METHODS
    new "classname", option => value, ...
        Create a YAML loader or dumper object with some options.

    SafeClass "classname", ...
        Register a string or list of strings to the list of allowed classes
        to the "Safe{Load,Dump}" methods. Without any SafeClass added, no
        custom "!" classes are allow in the YAML. Regexp are not supported.

    SafeLoad
    SafeLoadFile
        Restrict the loader to the registered safe classes only or tags
        starting with "perl/".

    SafeDump
    SafeDumpFile
        Restrict the dumper to the registered safe classes only or tags
        starting with "perl/".

    And all the loader and dumper options as getter and setter methods. See
    below.

Configuration
  Options for Loader and Dumper
    via getter and setter methods.

    "enablecode" =item "get_enablecode"
        If enabled turns on handling of code blocks for the loader and
        dumper. It sets both the "loadcode" and "dumpcode" option.

    "disableblessed" =item "get_disableblessed"
        Default 0

        If set, i.e. blessed is disabled, all blessed data for the Loader
        and Dumper is unblessed. This is similar to the Safe methods without
        any SafeClass set.

    "encoding" =item "get_encoding"
        Default "utf8"

        Set to "any", "utf8", "utf16le" or "utf16be".

    "boolean" =item "get_boolean"
        Set to "JSON::PP" or "boolean" to enable or 0 to disable. Encodes
        true and false to the respective classes. It will try to load
        JSON::PP or boolean and die if it can't be loaded.

        With that it's possible to add new "real" booleans to a data
        structure:

              my $o = YAML::Safe->new->boolean("JSON::PP"); # or "boolean"
              my $data = $o->Load("booltrue: true");
              $data->{boolfalse} = JSON::PP::false;
              my $yaml = Dump($data);
              # boolfalse: false
              # booltrue: true

        Please note that JSON::PP::Boolean and boolean.pm behave a bit
        differently. Ideally you should only use them in boolean context.
        Setting a boolean() class is only possible on a perl since v5.8.9.
        It will die on older perls.

        If not set, booleans are loaded as special perl variables
        "PL_sv_yes" and "PL_sv_no", which have the disadvantage that they
        are readonly, and you can't add those to an existing data structure
        with pure perl.

        If you simply need to load "perl booleans" that are true or false in
        boolean context, you will be fine with the default setting.

    "safemode" =item "get_safemode"
        Default 0

        If enabled by using the the Safe methods restrict the blessing only
        for the set of registered classes or tags starting with "perl/".

  Loader Options
    via getter and setter methods.

    "nonstrict" =item "get_nonstrict"
        If enabled permits certain reader errors to loosely match other YAML
        module semantics. In detail: Allow "control characters are not
        allowed" with while parsing a quoted scalar found unknown escape
        character. Note that any error is stored and returned, just not
        immediately. This is needed for cpan distroprefs.

        However the reader error "invalid trailing UTF-8 octet" and all
        other utf8 strictness violations are still fatal.

        And if the structure of the YAML document cannot be parsed, i.e. a
        required value consists only of invalid control characters, the
        loader returns an error, unlike with non-strict YAML modules.

    "loadcode" =item "get_loadcode"
        Turns on deparsing and evaling of code blocks in the loader.

  Dumper Options
    via globals variables or as optional getter and setter methods.

    "dumpcode" =item "get_dumpcode"
        If enabled supports Dump of CV code blocks via
        "YAML::Safe::coderef2text()".

    "quotenum" =item "get_quotenum"
        Default: enabled.

        If enabled strings that look like numbers but have not been numified
        will be quoted when dumping. This ensures leading that things like
        leading zeros and other formatting are preserved.

    "noindentmap" =item "get_noindentmap"
        If enabled fallback to the old "YAML::Safe" behavior to omit the
        indentation of map keys, which arguably violates the YAML spec, is
        different to all other YAML libraries and causes "YAML.pm" to fail.

        Disabled

             authors:
               - this author

        Enabled

             authors:
             - this author

    "indent" =item "get_indent"
        Default 2. Valid values are from 1 - 10.

    "wrapwidth" =item "get_wrapwidth"
        Default 80

        Control text wrapping.

    "canonical" =item "get_canonical"
        Default: disabled.

        Enable to sort map keys.

    "unicode" =item "get_unicode"
        Default 1

        Set to undef or 0 to disallow unescaped non-ASCII characters. e.g.
        "YAML::Safe-"new->unicode(0)>

    "linebreak" =item "get_linebreak"
        Default ln

        Set to "any", "cr", "ln" or "crln".

    "openended" =item "get_openended"
        Default 0

        If enabled embed the yaml into "...", if an explicit document end is
        required.

Using YAML::Safe with Unicode
    Handling unicode properly in Perl can be a pain. YAML::Safe only deals
    with streams of utf8 octets. Just remember this:

        $perl = Load($utf8_octets);
        $utf8_octets = Dump($perl);

    There are many, many places where things can go wrong with unicode. If
    you are having problems, use Devel::Peek on all the possible data
    points.

See Also
    YAML.pm
    YAML::XS
    YAML::Syck
    YAML::Tiny
    CPAN::Meta::YAML

Author
    Reini Urban <rurban@cpan.org>, based on YAML::XS by Ingy döt Net
    <ingy@cpan.org>

Copyright and License
    Copyright 2007-2016. Ingy döt Net. Copyright 2015-2019. Reini Urban.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://www.perl.com/perl/misc/Artistic.html>



Changes
---
version: 0.79_01
date:    Mon 17 Jun 2019 10:14:30 AM CEST
changes:
- Merge with cperl fixes (rurban) to be usable for CPAN and in core.
---
version: 0.79
date:    Tue 11 Jun 2019 09:35:22 PM CEST
changes:
- Support aliasing scalars resolved as null or booleans (tinita)
- Add YAML::XS::LibYAML::libyaml_version() (tinita)
- Support standard !!int/!!float tags instead of dying (tinita)
---
version: 0.78
date:    Sat 18 May 2019 06:35:21 PM CEST
changes:
- Fix double free/core dump when Dump()ing binary data, see issue#91 (tinita)
- Update config.h from libyaml (tinita)
---
version: 0.77
date:    Mon 15 Apr 2019 10:55:13 PM CEST
changes:
- "Update libyaml to version 0.2.2. Most important change for users is that
  plain urls in flow style can be parsed now. Example: `[ http://yaml.org ]`.
  (tinita)"
---
version: 0.76
date:    Sun Dec 30 20:10:16 CET 2018
changes:
- Add $Indent - number of spaces when dumping. PR#89 @zdm++
- Fix typo in docs. PR#87 @mr-mixas++
- Fix links in docs. @manwar++
---
version: 0.75
date:    Sat Nov  3 14:17:28 CET 2018
changes:
- Implement $LoadCode (tinita)
---
version: 0.74
date:    Fri Aug 31 18:06:17 PDT 2018
changes:
- Fix tests on older perls
- Support back to perl.5.8.1
---
version: 0.72
date:    Sun Jul  8 17:03:08 CEST 2018
changes:
- Update to libyaml 0.2.1 (tinita) It's forbidden now to escape single quotes
  inside double quotes
- When disabling $LoadBlessed, return scalars not refs (tinita)
- Save anchors also for blessed scalars (tinita)
---
version: 0.70
date:    Sat Jun  9 21:53:05 CEST 2018
changes:
- Fix format specifier/argument mismatch PR#79 @a3f++
- 'Travis CI: Test on Perl 5.26 PR#80 @a3f++'
- Fix a C90-compatibility issue PR#81 @stphnlyd++
---
version: 0.69
date:    Tue Dec 26 18:37:25 CET 2017
changes:
- Security fix Add $LoadBlessed option to turn on/off loading objects
  PR#73 PR#74 TINITA++
---
version: 0.68
date:    Mon Dec 18 20:01:05 CET 2017
changes:
- Fix regex roundtrip (fixes issue#69) PR#70 (TINITA++)
- Fix loading of many regexes (fixes issue#64) PR#71 (TINITA++)
---
version: 0.67
date:    Wed Nov 15 18:59:33 CET 2017
changes:
- Support standard tags !!str, !!map and !!seq instead of dying.
  PR#67 (TINITA++)
- Support JSON::PP::Boolean and boolean.pm via $YAML::XS::Boolean.
  PR#66 (TINITA++) Thanks also to Björn Höhrmann for SvPV_nolen()
---
version: 0.66
date:    Thu Aug 17 20:06:06 CEST 2017
changes:
- Apply PR#55 (@vti++)
- Dump() was modifying original data, adding a PV to numbers
- Added a test for this (TINITA++)
---
version: 0.65
date:    Thu 18 May 23:09:38 CEST 2017
changes:
- Prevent warning about unused variables (see issue#59)
- Clarify documentation about exported functions
---
version: 0.64
date:    Fri  7 Apr 20:32:01 CEST 2017
changes:
- use lib FindBin::Bin in tests, preparing for perl 5.26 where '.'
  gets removed from @INC. Fixes issue#54 TINITA++
- Use the latest libyaml sources.
- Lazy load B::Deparse for faster startup time.
  Applied PR/53 by @perlpunk++
  (submitted by @carnil++ from a Debian issue as GitHub issue#52)
---
version: 0.63
date:    Fri Jul  8 14:40:35 UTC 2016
changes:
- Apply PR/48 by @perlpunk++
- Fixes memory leaks
---
version: 0.62
date:    Mon Feb 22 15:46:06 UTC 2016
changes:
- Apply PR/42 @perlpunk++ (fix for detecting filehandles)
---
version: 0.61
date:    Sat Feb 20 18:04:19 UTC 2016
changes:
- Apply PR/37 @perlpunk++ (includes PR/36 @lameventanas++)
---
version: 0.60
date:    Tue Feb  9 14:36:09 EST 2016
changes:
- Apply PR/33. @andygrundman++. Solaris fixes.
---
version: 0.59
date:    Mon Jan 26 15:04:10 PST 2015
changes:
- PR/23 Better scalar dump heuristics
- More closely match YAML.pm
- Thanks Matthias Bethke
---
version: 0.58
date:    Tue Jan 20 21:01:49 PST 2015
changes:
- Add a VERSION statement to YAML::LibYAML (issue#8)
---
version: 0.57
date:    Thu Jan 15 23:05:15 EST 2015
changes:
- Applied fix for PR/21. nawglan++
---
version: 0.56
date:    Thu Jan 15 22:21:47 EST 2015
changes:
- Update copyright year
- Use Swim cpan-tail block functions in doc
---
version: 0.55
date:    Mon Dec 22 17:26:27 PST 2014
changes:
- Get YAML::XS using latest libyaml
---
version: 0.54
date:    Sat Nov 29 11:47:10 PST 2014
changes:
- Reverted an old patch with 0.53. Unreverted.
---
version: 0.53
date:    Fri Nov 28 09:20:08 PST 2014
changes:
- Fix for https://bitbucket.org/xi/libyaml/issue/10/wrapped-strings-cause-assert-failure
---
version: 0.52
date:    Fri Aug 22 21:03:19 PDT 2014
changes:
- Fix e1 test failure on 5.21.4
---
version: 0.51
date:    Sat Aug 16 14:28:45 PDT 2014
changes:
- Remove =travis section
---
version: 0.50
date:    Sat Aug 16 12:57:04 PDT 2014
changes:
- Meta 0.0.2
---
version: 0.49
date:    Sat Aug 16 04:23:17 PDT 2014
changes:
- Eliminate spurious trailing whitespace
---
version: 0.48
date:    Fri Aug 15 21:07:12 PDT 2014
changes:
- Add t/000-compile-modules.t
---
version: 0.47
date:    Sat Aug  9 00:29:29 PDT 2014
changes:
- Fix swim errors
---
version: 0.46
date:    Tue Aug  5 10:32:41 PDT 2014
changes:
- Add badges to doc
- Fix ReadMe
---
version: 0.45
date:    Mon Aug  4 01:23:10 PDT 2014
changes:
- Fix Meta and add Contributing.
---
version: 0.44
date:    Sun Jul 13 15:19:41 PDT 2014
changes:
- Doc fix. GitHub-Issue-#6. Thanks to Debian Perl Group for finding this.
---
version: 0.43
date:    Sat Jul 12 10:04:07 PDT 2014
changes:
- Test::Base tests needed 'inc' in @INC
---
version: 0.42
date:    Fri Jul 11 14:45:58 PDT 2014
changes:
- Switch to Zilla::Dist
- No longer dep on Test::Base, Spiffy, and Filter::Util::Call
- Remove test/changes.t
---
version: 0.41
date:    Wed Mar 13 10:34:55 PDT 2013
changes:
- Removed another C++ // style comment. jdb++
---
version: 0.40
date:    Tue Mar 12 11:05:34 PDT 2013
changes:
- Removed C++ // style comments, for better portability. jdb++
---
version: 0.39
date:    Mon Feb 11 18:08:04 PST 2013
changes:
- Using the latest libyaml codebase
- https://github.com/yaml/libyaml/tree/perl-yaml-xs

Safe.pm
use strict; use warnings;

package YAML::Safe;
our $VERSION = '0.80';
our $XS_VERSION = $VERSION;
$VERSION = eval $VERSION;

use base 'Exporter';
@YAML::Safe::EXPORT = qw(Load Dump);
@YAML::Safe::EXPORT_OK = qw(LoadFile DumpFile);
%YAML::Safe::EXPORT_TAGS = (
    all => [qw(Dump Load LoadFile DumpFile)],
);

use XSLoader;

#use Scalar::Util qw/ openhandle /;
#sub _DumpFile {
#    my $OUT;
#    my $filename = shift;
#    if (openhandle $filename) {
#        $OUT = $filename;
#    }
#    else {
#        my $mode = '>';
#        if ($filename =~ /^\s*(>{1,2})\s*(.*)$/) {
#            ($mode, $filename) = ($1, $2);
#        }
#        open $OUT, $mode, $filename
#          or die "Can't open '$filename' for output:\n$!";
#    }
#    local $/ = "\n"; # reset special to "sane"
#    print $OUT YAML::Safe::Dump(@_);
#}
#
#sub _LoadFile {
#    my $IN;
#    my $filename = shift;
#    if (openhandle $filename) {
#        $IN = $filename;
#    }
#    else {
#        open $IN, $filename
#          or die "Can't open '$filename' for input:\n$!";
#    }
#    return YAML::Safe::Load(do { local $/; local $_ = <$IN> });
#}

# XXX The following code should be moved from Perl to C.
$YAML::Safe::coderef2text = sub {
    my $coderef = shift;
    require B::Deparse;
    my $deparse = B::Deparse->new();
    my $text;
    eval {
        local $^W = 0;
        $text = $deparse->coderef2text($coderef);
    };
    if ($@) {
        warn "YAML::Safe failed to dump code ref:\n$@";
        return;
    }
    $text =~ s[BEGIN \{\$\{\^WARNING_BITS\} = "UUUUUUUUUUUU\\001"\}]
              [use warnings;]g;

    return $text;
};

$YAML::Safe::glob2hash = sub {
    my $hash = {};
    for my $type (qw(PACKAGE NAME SCALAR ARRAY HASH CODE IO)) {
        my $value = *{$_[0]}{$type};
        $value = $$value if $type eq 'SCALAR';
        if (defined $value) {
            if ($type eq 'IO') {
                my @stats = qw(device inode mode links uid gid rdev size
                               atime mtime ctime blksize blocks);
                undef $value;
                $value->{stat} = {};
                map {$value->{stat}{shift @stats} = $_} stat(*{$_[0]});
                $value->{fileno} = fileno(*{$_[0]});
                {
                    local $^W;
                    $value->{tell} = tell(*{$_[0]});
                }
            }
            $hash->{$type} = $value;
        }
    }
    return $hash;
};

use constant _QR_MAP => {
    '' => sub { qr{$_[0]} },
    x => sub { qr{$_[0]}x },
    i => sub { qr{$_[0]}i },
    s => sub { qr{$_[0]}s },
    m => sub { qr{$_[0]}m },
    ix => sub { qr{$_[0]}ix },
    sx => sub { qr{$_[0]}sx },
    mx => sub { qr{$_[0]}mx },
    si => sub { qr{$_[0]}si },
    mi => sub { qr{$_[0]}mi },
    ms => sub { qr{$_[0]}sm },
    six => sub { qr{$_[0]}six },
    mix => sub { qr{$_[0]}mix },
    msx => sub { qr{$_[0]}msx },
    msi => sub { qr{$_[0]}msi },
    msix => sub { qr{$_[0]}msix },
};

sub __qr_loader {
    if ($_[0] =~ /\A  \(\?  ([\^uixsm]*)  (?:-  (?:[ixsm]*))?  : (.*) \)  \z/x) {
        my ($flags, $re) = ($1, $2);
        $flags =~ s/^\^//;
        $flags =~ tr/u//d;
        my $sub = _QR_MAP->{$flags} || _QR_MAP->{''};
        my $qr = &$sub($re);
        return $qr;
    }
    return qr/$_[0]/;
}

sub __code_loader {
    my ($string) = @_;
    my $sub = eval "sub $string";
    if ($@) {
        warn "YAML::Safe failed to load sub: $@";
        return sub {};
    }
    return $sub;
}

XSLoader::load 'YAML::Safe', $XS_VERSION;

1;
__END__
=encoding UTF-8

=head1 Name

YAML::Safe - Safe Perl YAML Serialization using XS and libyaml

=for html
<a href="https://travis-ci.org/rurban/YAML-Safe"><img src="https://travis-ci.org/rurban/YAML-Safe.png" alt="YAML-Safe"></a>

=head1 Synopsis

    use YAML::Safe qw(LoadFile DumpFile);

    my $yaml  = Dump [ 1..4 ];
    my $array = Load $yaml;

    $yaml  = DumpFile ("my.yml", [ 1..4 ]);
    $array = LoadFile "my.yml";
    open my $fh, "my.yml";
    $yaml  = DumpFile ($fh, [ 1..4 ]);
    $array = LoadFile $fh;
    open *FH, "my.yml";
    $yaml  = DumpFile (*FH, [ 1..4 ]);
    $array = LoadFile *FH;

    my $obj = YAML::Safe->new;
    $yaml  = $obj->DumpFile ("my.yml", [ 1..4 ]);
    $array = $obj->LoadFile("my.yml");

    $yaml = YAML::Safe->new->nonstrict->encoding("any");
    $yaml->SafeClass("DateTime");
    $array = $yaml->SafeLoadFile("META.yml");

    $yaml = YAML::Safe->new->canonical;
    $yaml->SafeClass("DateTime");
    $array = $yaml->SafeDumpFile("META.yml");

=head1 Description

This module is a refactoring of L<YAML::XS>, the old Perl XS binding
to C<libyaml> which offers Perl somewhat acceptable YAML support to
date.  YAML::XS never produced code which could be read from YAML, and
thus was unsuitable to be used as YAML replacement for core and CPAN.
It also required reading and setting options from global variables.

Kirill Siminov's C<libyaml> is a good YAML library implementation. The
C library is written precisely to the YAML 1.1 specification, and
offers YAML 1.2 support. It was originally bound to Python and was
later bound to Ruby.  C<libsyck> is written a bit more elegant, has
less bugs, is not as strict as libyaml, but misses some YAML
features. It can only do YAML 1.0

This module exports the functions C<Dump> and C<Load>, and do work as
functions exactly like L<YAML::XS> and C<YAML.pm>'s corresponding
functions.  It is however preferred to use the new OO-interface to
store all options in the new created object. YAML::Safe does not
support the old globals anymore.

There are also new Safe variants of Load and Dump methods, and
options as setter methods.

If you set the option C<noindentmap>, C<YAML::Safe> will behave like
with version E<lt> 0.70, which creates yml files which cannot be read
by C<YAML.pm>

However the loader is stricter than C<YAML>, C<YAML::Syck> and
C<CPAN::Meta::YAML> i.e. C<YAML::Tiny> as used in core. Set the option
C<nonstrict> to allow certain reader errors to pass the
C<CPAN::Meta> validation testsuite.

=head1 FUNCTIONS

=over

=item Load

=item LoadFile

=item Dump

=item DumpFile

=item libyaml_version

=back

=head1 METHODS

=over

=item new "classname", option => value, ...

Create a YAML loader or dumper object with some options.

=item SafeClass "classname", ...

Register a string or list of strings to the list of allowed classes to
the C<Safe{Load,Dump}> methods. Without any SafeClass added, no custom
C<!> classes are allow in the YAML.  Regexp are not supported.

=item SafeLoad

=item SafeLoadFile

Restrict the loader to the registered safe classes only
or tags starting with "perl/".

=item SafeDump

=item SafeDumpFile

Restrict the dumper to the registered safe classes only
or tags starting with "perl/".

=back

And all the loader and dumper options as getter and setter methods.
See below.

=head1 Configuration

=head2 Options for Loader and Dumper

via getter and setter methods.

=over

=item C<enablecode>
=item C<get_enablecode>

If enabled turns on handling of code blocks for the loader and dumper.
It sets both the C<loadcode> and C<dumpcode> option.

=item C<disableblessed>
=item C<get_disableblessed>

Default 0

If set, i.e. blessed is disabled, all blessed data for the Loader and
Dumper is unblessed.  This is similar to the Safe methods without any
SafeClass set.

=item C<encoding>
=item C<get_encoding>

Default "utf8"

Set to "any", "utf8", "utf16le" or "utf16be".

=item C<boolean>
=item C<get_boolean>

Set to "JSON::PP" or "boolean" to enable or 0 to disable.  Encodes
true and false to the respective classes. It will try to load
L<JSON::PP> or L<boolean> and die if it can't be loaded.

With that it's possible to add new "real" booleans to a data structure:

      my $o = YAML::Safe->new->boolean("JSON::PP"); # or "boolean"
      my $data = $o->Load("booltrue: true");
      $data->{boolfalse} = JSON::PP::false;
      my $yaml = Dump($data);
      # boolfalse: false
      # booltrue: true

Please note that JSON::PP::Boolean and boolean.pm behave a bit differently.
Ideally you should only use them in boolean context.
Setting a boolean() class is only possible on a perl since v5.8.9. It will die
on older perls.

If not set, booleans are loaded as special perl variables C<PL_sv_yes> and
C<PL_sv_no>, which have the disadvantage that they are readonly, and you can't
add those to an existing data structure with pure perl.

If you simply need to load "perl booleans" that are true or false in boolean
context, you will be fine with the default setting.

=item C<safemode>
=item C<get_safemode>

Default 0

If enabled by using the the Safe methods restrict the blessing only for
the set of registered classes or tags starting with "perl/".

=back

=head2 Loader Options

via getter and setter methods.

=over

=item C<nonstrict>
=item C<get_nonstrict>

If enabled permits certain reader errors to loosely match other YAML
module semantics. In detail: Allow B<"control characters are not
allowed"> with while parsing a quoted scalar found unknown escape
character. Note that any error is stored and returned, just not
immediately. This is needed for cpan distroprefs.

However the reader error B<"invalid trailing UTF-8 octet"> and all
other utf8 strictness violations are still fatal.

And if the structure of the YAML document cannot be parsed, i.e. a
required value consists only of invalid control characters, the loader
returns an error, unlike with non-strict YAML modules.

=item C<loadcode>
=item C<get_loadcode>

Turns on deparsing and evaling of code blocks in the loader.

=back

=head2 Dumper Options

via globals variables or as optional getter and setter methods.

=over

=item C<dumpcode>
=item C<get_dumpcode>

If enabled supports Dump of CV code blocks via
C<YAML::Safe::coderef2text()>.

=item C<quotenum>
=item C<get_quotenum>

Default: enabled.

If enabled strings that look like numbers but have not
been numified will be quoted when dumping.
This ensures leading that things like leading zeros and other
formatting are preserved.

=item C<noindentmap>
=item C<get_noindentmap>

If enabled fallback to the old C<YAML::Safe> behavior to omit the
indentation of map keys, which arguably violates the YAML spec, is
different to all other YAML libraries and causes C<YAML.pm> to fail.

Disabled

     authors:
       - this author

Enabled

     authors:
     - this author

=item C<indent>
=item C<get_indent>

Default 2.
Valid values are from 1 - 10.

=item C<wrapwidth>
=item C<get_wrapwidth>

Default 80

Control text wrapping.

=item C<canonical>
=item C<get_canonical>

Default: disabled.

Enable to sort map keys.

=item C<unicode>
=item C<get_unicode>

Default 1

Set to undef or 0 to disallow unescaped non-ASCII characters.
e.g. C<YAML::Safe->new->unicode(0)>

=item C<linebreak>
=item C<get_linebreak>

Default ln

Set to "any", "cr", "ln" or "crln".

=item C<openended>
=item C<get_openended>

Default 0

If enabled embed the yaml into "...", if an explicit
document end is required.

=back

=head1 Using YAML::Safe with Unicode

Handling unicode properly in Perl can be a pain. YAML::Safe only deals
with streams of utf8 octets. Just remember this:

    $perl = Load($utf8_octets);
    $utf8_octets = Dump($perl);

There are many, many places where things can go wrong with unicode. If
you are having problems, use Devel::Peek on all the possible data
points.

=head1 See Also

=over

=item L<YAML>.pm

=item L<YAML::XS>

=item L<YAML::Syck>

=item L<YAML::Tiny>

=item L<CPAN::Meta::YAML>

=back

=head1 Author

Reini Urban <rurban@cpan.org>, 
based on YAML::XS by Ingy döt Net <ingy@cpan.org>

=head1 Copyright and License

Copyright 2007-2016. Ingy döt Net.
Copyright 2015-2019. Reini Urban.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut

